Tools/clinic/clinic.py

```diff
@@ -4840,6 +4830,22 @@ def state_dsl_start(self, line: str) -> None:
 
         self.next(self.state_modulename_name, line)
 
+    @staticmethod
+    def parse_names(line: str) -> tuple[str, str]:
+        left, _, right = line.partition(' as ')
+        full_name = left.strip()
+        c_basename = right.strip()
+        if not c_basename:
+            fields = full_name.split(".")
+            if fields[-1] == '__new__':
+                fields.pop()
+            c_basename = "_".join(fields)
+        if not is_legal_py_identifier(full_name):
+            fail(f"Illegal function name: {full_name!r}")
+        if not is_legal_c_identifier(c_basename):
```

Comment:
In both call sites currently, the code is careful not to `fail()` if `right` is an empty string or a string consisting only of whitespace. Should we do the same in this refactor?

```suggestion
        if c_basename and not is_legal_c_identifier(c_basename):
```

Reply:
Before, `c_basename` could be `None` all the way up to the renderer. Now, we ensure it is a valid C basename right from the start, so we don't need this guard anymore.

Reply:
Ahh, but the rewrite does something different, because you now have the extra block above:

```py
        if not c_basename:
            fields = full_name.split(".")
            if fields[-1] == '__new__':
                fields.pop()
            c_basename = "_".join(fields)
```

Sorry, I should have seen this.